pkg/service/service.go

```diff
@@ -247,15 +249,19 @@ func (s *Service) UpsertService(params *UpsertServiceParams) (bool, lb.ID, error
 
 	backendsCopy := []lb.Backend{}
 	for _, b := range params.Backends {
-		// Services with trafficPolicy=Local may only use node-local backends
-		// for external scope. We implement this by filtering out all backend
-		// IPs which are not a local endpoint.
+		// Local redirect services or services with trafficPolicy=Local may
+		// only use node-local backends for external scope. We implement this by
+		// filtering out all backend IPs which are not a local endpoint.
 		if filterBackends && len(b.NodeName) > 0 && b.NodeName != nodeTypes.GetName() {
 			continue
 		}
 		backendsCopy = append(backendsCopy, *b.DeepCopy())
 	}
 
+	// TODO (Aditi) When we filter backends for LocalRedirect service, there
+	// might be some backend pods with active connections. We may need to
+	// defer filtering the backends list (thereby defer redirecting traffic)
+	// in such cases.
```

Comment:
Which case are you thinking of here? Is it a service that already exists with various backends but is then configured to become a local redirect, hence should only send to local backends?

Reply:
Yes, when a user configures an LRP with Kubernetes service as `redirectFrom`, the service may have a mix of local + remote backend with active connections. We may need to hook into conntrack information to learn about the existing active connections. I need to think more about this case. I'll address this in a follow-up PR.

Reply:
:+1: for follow-up, makes sense. Also for the record, this was discussed during 2020-08-12 SIG-datapath (if someone wants to look it up).

Reply:
I think all this needs a clear documentation somewhere as part of this PR with regards to semantics that users need to expect when configuring this. We were also discussing during SIG-datapath that existing socket LB service connections for connected TCP/UDP doesn't cause the local redirect to take effect. I'm not quite sure yet how that could be addressed.

Reply:
Agreed, I plan to document all the caveats. I didn't follow your 2nd comment. The redirection will take effect as soon as we override LB map entries. Per my TODO note above, we probably need to consider cases where there are active connections to remote pods, and we may need to defer applying the local redirection in the datapath. As discussed in the meeting, since the current use cases (kiam and node-local dns) are likely to have short connections so the probability of breaking those would be low.

Reply:
CT deletion is possible but that wouldn't address the socket LB case from Daniel above. At least from my perspective this limitation is documented and not necessary for initial implementation.